rustls/src/server/hs.rs

```diff
@@ -201,6 +206,83 @@ impl ExtensionProcessing {
     }
+
+    fn validate_server_cert_type_extension(
+        &mut self,
+        hello: &ClientHelloPayload,
+        config: &ServerConfig,
+        cx: &mut ServerContext<'_>,
+    ) -> Result<(), Error> {
+        let requires_server_rpk = config
+            .cert_resolver
+            .only_raw_public_keys();
+        let client_allows_rpk = hello
+            .server_certificate_extension()
+            .map(|certificate_types| certificate_types.contains(&CertificateType::RawPublicKey))
+            .unwrap_or(false);
+        let raw_key_negotation = cx
+            .common
+            .validate_rpk_negotiation(RawKeyNegotiationParams {
+                peer_supports_raw_key: client_allows_rpk,
+                local_expects_raw_key: requires_server_rpk,
+                extension_type: ExtensionType::ServerCertificateType,
+            });
+        self.process_cert_type_extension(raw_key_negotation, cx)
+    }
+
+    fn validate_client_cert_type_extension(
+        &mut self,
+        hello: &ClientHelloPayload,
+        config: &ServerConfig,
+        cx: &mut ServerContext<'_>,
+    ) -> Result<(), Error> {
+        let requires_client_rpk = config
+            .verifier
+            .requires_raw_public_keys();
+        let client_offers_rpk = hello
+            .client_certificate_extension()
+            .map(|certificate_types| certificate_types.contains(&CertificateType::RawPublicKey))
+            .unwrap_or(false);
+        let raw_key_negotation = cx
+            .common
+            .validate_rpk_negotiation(RawKeyNegotiationParams {
+                peer_supports_raw_key: client_offers_rpk,
+                local_expects_raw_key: requires_client_rpk,
+                extension_type: ExtensionType::ClientCertificateType,
+            });
+        self.process_cert_type_extension(raw_key_negotation, cx)
+    }
+
+    fn process_cert_type_extension(
+        &mut self,
+        raw_key_negotiation_result: RawKeyNegotationResult,
+        cx: &mut ServerContext<'_>,
+    ) -> Result<(), Error> {
+        match raw_key_negotiation_result {
+            RawKeyNegotationResult::Negotiated(ExtensionType::ClientCertificateType) => {
+                self.exts
+                    .push(ServerExtension::ClientCertType(
+                        CertificateType::RawPublicKey,
+                    ));
+            }
+            RawKeyNegotationResult::Negotiated(ExtensionType::ServerCertificateType) => {
+                self.exts
+                    .push(ServerExtension::ServerCertType(
+                        CertificateType::RawPublicKey,
+                    ));
+            }
+            RawKeyNegotationResult::Err(err) => {
+                return Err(cx
+                    .common
+                    .send_fatal_alert(AlertDescription::HandshakeFailure, err));
+            }
+            RawKeyNegotationResult::NotNegotiated => {}
+            RawKeyNegotationResult::Negotiated(_) => {
+                unreachable!();
```

Comment:
Nit: remove the `;` to make this more compact. Maybe add a comment on the arm why this is unreachable?